Add rails db:system:change command

gmcgibbon opened the pull request:
Part two of https://github.com/rails/rails/issues/34710.

After studying up commands and generators, I decided to make a command that delegates to a generator (similar to rails new / app generator). This seems to give us the most flexibility with how the command interacts with files, and if the user wants to accept the config overwrites or not. 

TODO:

- [x] Fix three level nested generator lookup
- [x] Add tests for generator with file assertions
- [x] Add chagelog entry

gmcgibbon committed:
Move application generator naming and database code to concerns

gmcgibbon committed:
Add rails db:system:change command

Add `rails db:system:change` command for changing databases.

```
bin/rails db:system:change --to=postgresql
   force  config/database.yml
    gsub  Gemfile
```

The change command copies a template `config/database.yml` with
the target database adapter into your app, and replaces your database
gem with the target database gem.

eileencodes commented on railties/lib/rails/generators/database.rb:
```diff
@@ -0,0 +1,60 @@
+        when "jdbcsqlite3"    then ["activerecord-jdbcsqlite3-adapter", nil]
+        when "jdbcpostgresql" then ["activerecord-jdbcpostgresql-adapter", nil]
+        when "jdbc"           then ["activerecord-jdbc-adapter", nil]
+        else [database, nil]
```
Should this raise an error if we get to the else? Otherwise I think this will silently fail.

eileencodes commented on railties/lib/rails/generators/rails/db/system/change/change_generator.rb:
```diff
@@ -0,0 +1,55 @@
+          def initialize(*)
+            super
+            if !DATABASES.include?(options[:to])
+              raise Error, "Invalid value for --to option. Supported for preconfiguration are: #{DATABASES.join(", ")}."
```
This sentence reads a little weird to me. I think "Supported preconfigurations are" would sound better than "Supported for preconfiguration are"

eileencodes reviewed (changes requested):
I like this. I left 2 minor comments.

gmcgibbon committed:
Revise wording on invalid database error messages

kaspth commented on railties/lib/rails/generators.rb:
```diff
@@ -239,8 +241,15 @@ def sorted_groups
+
+        if context
+          lookups << "#{name}:#{context}"
+        end
```
What do these new lines do? I've always found the `base` and `context` naming here confusing?

kaspth replied:
So it's possible for users to call rails g db:system:change? I thought this generator was internal and sufficiently called via it's class start method from the command?

gmcgibbon replied:
I'll go ahead and revert the changes to the generator lookup and hide this generator so it can only be called through the command. I'm assuming we would prefer only one way of changing databases.

kaspth commented on railties/lib/rails/commands/db/system/change/change_command.rb:
```diff
@@ -0,0 +1,20 @@
+    module Db
+      module System
+        class ChangeCommand < Base # :nodoc:
+          class_option :to, desc: "The DBMS to switch to."
```
Do we generally say DBMS elsewhere?

matthewd replied:
Try "database type"? It's not brilliant, but maybe an improvement? Or "database platform"?

kaspth replied:
Using "database system" would match the command name and reinforce the concept. But it's perhaps not expounding on what a "database system" means, that meaning only comes from seeing MySQL, Postgres and the gang.

gmcgibbon replied:
Agree that `database system` makes the most sense given the command name.

kaspth commented on railties/lib/rails/generators/app_name.rb:
```diff
@@ -0,0 +1,51 @@
+      RESERVED_NAMES = %w(application destroy plugin runner test)
+
+      private
+
```
✂️

kaspth commented on railties/lib/rails/generators/app_name.rb:
```diff
@@ -0,0 +1,51 @@
+
+        def defined_app_const_base
+          Rails.respond_to?(:application) && defined?(Rails::Application) &&
+            Rails.application.is_a?(Rails::Application) && Rails.application.class.name.sub(/::Application$/, "")
```
Could switch to `chomp` instead of `sub` while we're here.

kaspth commented on railties/lib/rails/generators/app_name.rb:
```diff
@@ -0,0 +1,51 @@
+
+        def valid_const?
+          if /^\d/.match?(app_const)
+            raise Error, "Invalid application name #{original_app_name}. Please give a name which does not start with numbers."
```
Curiosity for another time: since 2.6 allows constants to start with non ascii characters, does that mean we could name a Rails app, say, 666?

jrochkind replied:
@kaspth not `666`, because it needs to begin with something the unicode database as included in the particular ruby version thinks is an uppercase letter.  Just not limited to ascii.

kaspth commented on railties/lib/rails/generators/database.rb:
```diff
@@ -0,0 +1,60 @@
+      end
+
+      def gem_for_database(database = options[:database])
+        # %w( mysql postgresql sqlite3 oracle frontbase ibm_db sqlserver jdbcmysql jdbcsqlite3 jdbcpostgresql )
```
What is this for?

kaspth commented on railties/lib/rails/generators/database.rb:
```diff
@@ -0,0 +1,60 @@
+    module Database # :nodoc:
+      DATABASES = %w( mysql postgresql sqlite3 oracle frontbase ibm_db sqlserver )
+      JDBC_DATABASES = %w( jdbcmysql jdbcsqlite3 jdbcpostgresql jdbc )
+      DATABASES.concat(JDBC_DATABASES)
```
Seems odd to mutate it right after defining it. Maybe swap definition order and use `+`?

kaspth commented on railties/lib/rails/generators/database.rb:
```diff
@@ -0,0 +1,60 @@
+      end
+
+      private
+
```
✂️

kaspth commented on railties/lib/rails/generators/rails/db/system/change/change_generator.rb:
```diff
@@ -0,0 +1,55 @@
+            end
+            self.options = options.dup
+            options["database"] ||= options["to"]
+            options.freeze
```
Perhaps we should mirror the options dup and freeze done in convert_database_option_for_jruby

kaspth commented on railties/lib/rails/generators/rails/db/system/change/change_generator.rb:
```diff
@@ -0,0 +1,55 @@
+            super
+            if !DATABASES.include?(options[:to])
+              raise Error, "Invalid value for --to option. Supported preconfigurations are: #{DATABASES.join(", ")}."
+            end
```
Add a newline before and after this `if`. Could also swap to `unless`.

kaspth commented on railties/lib/rails/generators/rails/db/system/change/change_generator.rb:
```diff
@@ -0,0 +1,55 @@
+          end
+
+          private
+
```
✂️

kaspth commented on railties/lib/rails/generators/rails/db/system/change/change_generator.rb:
```diff
@@ -0,0 +1,55 @@
+            end
+
+            def all_database_gems_regex
+              database_gem_name, _ = gem_for_database
```
Not used?

kaspth replied:
I should clarify: `all_database_gems_regex` doesn't use `database_gem_name, _ =` that you assign. `edit_gemfile` does however.

gmcgibbon replied:
🤦‍♂️ Thanks, I'll remove it.

kaspth commented on railties/lib/rails/generators/rails/db/system/change/change_generator.rb:
```diff
@@ -0,0 +1,55 @@
+
+            def all_database_gems_regex
+              database_gem_name, _ = gem_for_database
+              all_database_gem_names = all_database_gems.map { |name, _| name }
```
`keys` instead of `map`, or perhaps `map(&:first)`? Should probably also be moved to `all_database_gems`.

kaspth commented on railties/test/generators/generators_test_helper.rb:
```diff
@@ -63,4 +63,28 @@ def copy_routes
+        gem 'turbolinks', '~> 5'
+        # Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
+        gem 'jbuilder', '~> 2.5'
+      GEMFILE
```
How will we have to maintain these Gemfiles going forward?

kaspth commented on railties/test/generators/db_system_change_generator_test.rb:
```diff
@@ -0,0 +1,67 @@
+              assert_match "adapter: postgresql", content
+              assert_match "database: test_app", content
+            end
+            assert_file("Gemfile") do |content|
```
Adding a newline after `output` and between the `assert_file` seems like it would increase readability in these tests.

kaspth commented on railties/lib/rails/commands/db/system/change/change_command.rb:
```diff
@@ -0,0 +1,20 @@
+  module Command
+    module Db
+      module System
+        class ChangeCommand < Base # :nodoc:
```
For another time: seeing this deep module nesting makes me think we should consider adding something akin to Action Mailboxes' `routing` for commands.

kaspth reviewed (commented):
You're getting pretty adept at finding your way around the codebase! Lovely to see 💪

kaspth commented:
Forgot to ask, how are we handling cases where people who may have changed their `config/database.yml` when calling this command?

gmcgibbon commented:
> Forgot to ask, how are we handling cases where people who may have changed their config/database.yml when calling this command?

The generator will detect a configuration file already exists and gives options to overwrite, show a diff, and reject the change.

gmcgibbon commented:
I've hidden the change generator, reverted the patch to call it with `rails g`, and removed the static gemfile strings in favour of template evaluation. Let me know if there's anything else this PR needs to get shipped!

kaspth committed:
Merge branch 'master' into db_system_change_command

bogdanvlviv commented:
Hey, Thanks for adding this!
Since this PR is merged we should close #34710, and #34732?
Also, I think we should add `db:change` to `COMMANDS_IN_USAGE`(`railties/lib/rails/command.rb` file) in order to not show it on `rails --help`, (we should show `db:system:change` instead).
Sorry for the too late review.

rafaelfranca commented:
cc @gmcgibbon

gmcgibbon commented:
Thanks for pointing that out @bogdanvlviv, I'll see what I can do!

simi commented:
Implementing this as a command makes it inconsistent at few places:

1. `bin/rails db:system:change --help` shows wrong usage (there is just `change` instead of `db:system:change`)

```
[retro@retro  test-app (master #%)]❤ bin/rails db:system:change --help
Usage:
  rails change [options]

Options:
  [--to=TO]  # The database system to switch to.

```

2. this command is not present for `bin/rake -T`, but all other `db:*` commands are
3. this command is not present for `bin/rails --tasks`, but all other `db:*` commands are

But it is present for `rails` command.

```
[retro@retro  test-app (master #%)]❤ bin/rails
The most common rails commands are:
 generate     Generate new code (short-cut alias: "g")
 console      Start the Rails console (short-cut alias: "c")
 server       Start the Rails server (short-cut alias: "s")
 test         Run tests except system tests (short-cut alias: "t")
 test:system  Run system tests
 dbconsole    Start a console for the database specified in config/database.yml
              (short-cut alias: "db")

 new          Create a new Rails application. "rails new my_app" creates a
              new application called MyApp in "./my_app"


All commands can be run with -h (or --help) for more information.
In addition to those commands, there are:

  about
  action_mailbox:ingress:exim
  action_mailbox:ingress:postfix
  action_mailbox:ingress:qmail
  action_mailbox:install
  action_text:install
  action_text:install:migrations
  active_storage:install
  app:template
  app:update
  assets:clean[keep]
  assets:clobber
  assets:environment
  assets:precompile
  cache_digests:dependencies
  cache_digests:nested_dependencies
  credentials:diff
  credentials:edit
  credentials:show
  db:create
  db:drop
  db:environment:set
  db:fixtures:load
  db:migrate
  db:migrate:down
  db:migrate:redo
  db:migrate:status
  db:migrate:up
  db:prepare
  db:reset
  db:rollback
  db:schema:cache:clear
  db:schema:cache:dump
  db:schema:dump
  db:schema:load
  db:seed
  db:seed:replant
  db:setup
  db:structure:dump
  db:structure:load
  db:system:change
  db:version
  destroy
  dev:cache
  encrypted:edit
  encrypted:show
  initializers
  log:clear
  middleware
  notes
  restart
  routes
  runner
  secret
  secrets:edit
  secrets:setup
  secrets:show
  stats
  test:db
  time:zones[country_or_offset]
  tmp:clear
  tmp:create
  version
  webpacker
  webpacker:binstubs
  webpacker:check_binstubs
  webpacker:check_node
  webpacker:check_yarn
  webpacker:clean[keep]
  webpacker:clobber
  webpacker:compile
  webpacker:info
  webpacker:install
  webpacker:install:angular
  webpacker:install:coffee
  webpacker:install:elm
  webpacker:install:erb
  webpacker:install:react
  webpacker:install:stimulus
  webpacker:install:svelte
  webpacker:install:typescript
  webpacker:install:vue
  webpacker:verify_install
  webpacker:yarn_install
  yarn:install
  zeitwerk:check

```
